### PR TITLE
Add gluon losses

### PR DESCRIPTION
**Pull request by piiswrong**
Added back some of the losses reverted in #8010

Losses that are not added back include
```python
class MultiClassHingeLoss(Loss):
    """Calculates the MaxMargin loss, aka multiclass soft-margin
       loss. This requires access to a multiclass loss matrix delta
       which measures the cost for misclassifying label y as y'. This
       matrix can be specified at construction time. If it does not
       exist, we will susbstitute it with a 0-1 loss with automagic
       size inference.

    .. math::
       L = {max}_{y} [\\delta({label}, y) + {output}[y]] - {output}_{label}

    Label's shape should be output's shape without the `axis` dimension. i.e. for
    `output.shape` = (1,2,3,4) and axis = 2, `label.shape` should be (1,2,4).

    Parameters
    ----------
    delta : loss matrix, default None. In this case it is presumed to
        be a (0,1) loss, i.e. a constant loss of 1 for all
        misclassifications. Otherwise its dimensionality must match
        that of the number of classes.
    axis : int, default -1
        The axis to sum over when taking the maximum.
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, delta=None, axis=-1, weight=None, batch_axis=0, **kwargs):
        super(MaxMargin, self).__init__(weight, batch_axis, **kwargs)
        self._axis = axis
        self._delta = delta
        super(MaxMargin, self).__init__(weight, batch_axis, **kwargs)

    def hybrid_forward(self, F, output, label, sample_weight=None):
        # Check if the cost matrix has been defined. If not, use a
        # dumb (0,1) loss. This is only executed once, the first time
        # you invoke the loss.
        if (self._delta is None):
            classes = output.shape[self._axis]
            self._delta = F.ones(shape=(classes, classes))
            for i in range(classes):
                self._delta[i, i] = 0
        loss = -F.pick(output, label, axis=self._axis, keepdims=True)
        loss += F.max(output + F.take(self._delta, label), axis=self._axis, keepdims=True)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)
```
there are multiple versions of mult hinge loss. Also the implementation here won't work on GPU. Use of delta matrix is over complicated.

```python
class EpsilonInsensitiveLoss(Loss):
    """Calculates a loss that is insensitive to absolute error smaller than
    epsilon.

    .. math::
        L = \\mathrm{max}(0, |{output}_i - {label}_i| - \\epsilon)

    Output and label must have the same number of elements.

    Parameters
    ----------
    epsilon : float
        Threshold for trimmed epsilon-insensitivity parameter. By default set to 0.1
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, epsilon, weight=None, batch_axis=0, **kwargs):
        super(EpsilonInsensitiveLoss, self).__init__(weight, batch_axis, **kwargs)
        self._epsilon = epsilon

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = F.relu(F.abs(output - label) - self._epsilon)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
Don't like the name. Is there a better one?

```python

class ExponentialLoss(Loss):
    """Calculates the exponential hinge loss (quite obscure):

    .. math::
        L = \\exp(- {output}_i {label}_i)

    Output and label must have the same shape.

    Parameters
    ----------
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, weight=None, batch_axis=0, **kwargs):
        super(Exponential, self).__init__(weight, batch_axis, **kwargs)

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = F.exp(-output * label)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
Since its obscure, do we still need to add it?

```python

class QuantileLoss(Loss):
    """Calculates Koenker's quantile regression loss function yielding an estimate of the
       appropriately chosen quantile rather than the mean (or median):

    .. math::
        L = {max}(\\tau ({output}_i - {label}_i), (1-\\tau) ({label}_i - {output}_i)

    Output and label must have the same shape.

    Parameters
    ----------
    tau : float
        Quantile of the estimator. By default set to 0.5, i.e. by
        default identical with L1 loss, up to a scaling factor.
        This must be in the range (0,1).
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, tau=0.5, weight=None, batch_axis=0, **kwargs):
        super(Quantile, self).__init__(weight, batch_axis, **kwargs)
        self._tau = tau

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = output - label
        loss = F.maximum(self._tau * loss, (self._tau - 1.0)* loss)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
What is this for?

```python
class LangfordLoss(Loss):
    """Calculates the Huberized soft-margin loss that is used in VW (Vowpal Wabbit).
       It is given by a squared loss for margin values of [-1, 0] and by a linear
       loss for values larger than that.

    .. math::
        L = \\begin{cases}
          0 &
          \\text{ if } {output}_i {label}_i > 1 \\\
          \\frac{1}{2} - {output}_i {label}_i &
          \\text{ if } {output}_i {label}_i < 0 \\\
          \\frac{1}{2} (1 - {output}_i {label}_i)^2 &
          \\text{ otherwise }
          \\end{cases}

    Output and label must have the same shape.

    Parameters
    ----------
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, weight=None, batch_axis=0, **kwargs):
        super(Langford, self).__init__(weight, batch_axis, **kwargs)

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = F.relu(1 - output * label)
        loss = F.where(loss < 1.0, 0.5 * (loss**2), loss - 0.5)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
add back later.

```python
class DualKLLoss(Loss):
    """Estimates the Kullback Leibler Divergence between two
       distributions by convex duality. See Nguyen, Wainwright and
       Jordan (NGW), 2008 for a detailed derivation. In a nutshell it
       estimates:

    .. math::
       KL(p\\|q) = E_p[\\log p(x)] - E_p[\\log q(x)]

       Clearly this isn't easy to compute. Hence, NGW use the dual of
       the F-divergence log p(x)/q(x) and pose it as an optimization
       problem. This leads to the following loss function, which is
       different for both distributions (which we treat as a binary
       classification problem). The function that is being estimated
       allows us to get the Radon-Nikodym via dp/dq = exp(f).

    .. math::
        L = \\begin{cases}
            \\exp(f) & \\text{ if } y = -1 \\\
             -f-1 & \text{ if } y = 1
          \\end{cases}

    Output and label must have the same shape.

    Parameters
    ----------
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, weight=None, batch_axis=0, **kwargs):
        super(DualKL, self).__init__(weight, batch_axis, **kwargs)

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = F.where(label, output + 1, F.exp(output))
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
Add back later.

```python
class RelativeNoveltyLoss(Loss):
    """Estimates a relative novelty detector. See the Song, Teo and
       Smola (STS), 2009 for details. The main point is to estimate
       the ratio dp/dq well via max(0, rho - log dp/dq). As with the
       KL divergence estimator, the Fenchel-Legendre dual is easier to
       deal with. This leads to the following loss function:

    .. math::
        L = \\begin{cases}
            \\exp(f - rho) & \\text{ if } y = -1 \\\
             -f-1 & \\text{ if } y = 1 \\text{ and } f > 0 \\
            \\exp(f) & \\text{ if } y = 1 \\text{ and } f <= 0
          \\end{cases}

    output and label must have the same shape.

    Parameters
    ----------
    rho : float
        Relative probability weight for the most prevalent part of the
        probability distribution. It needs to be (0, 1).
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, rho=0.1, weight=None, batch_axis=0, **kwargs):
        super(RelativeNovelty, self).__init__(weight, batch_axis, **kwargs)
        self._rho = rho

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = - F.where(output > 0, output + 1,  F.exp(output))
        loss = F.where(label, loss, F.exp(output - self._rho))
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
a better name?

```python
class LogCoshLoss(Loss):
    """Calculates the smoothed L1 loss, aka log cosh loss in a
       numerically stable manner (i.e. without exponentiating large
       values of the cosh function.

    .. math::
        L = \\log 2 \\cosh ({output}_i - {label}_i)

    Output and label must have the same shape.

    Parameters
    ----------
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, weight=None, batch_axis=0, **kwargs):
        super(LogCosh, self).__init__(weight, batch_axis, **kwargs)

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = F.abs(label - output)
        loss = loss + F.log(0.5 + 0.5 * F.exp(-2 * loss))
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)

```
better name?

```python
class PoissonLoss(Loss):
    """Calculates the Poisson loss function (up to the normalization
       by a factorial in the label, due to computational efficiency
       reasons).

       NOTE THAT THIS IS DIFFERENT FROM THE POISSON LOSS IN PYTORCH
       AND KERAS INSOFAR AS IT USES THE EPXONENTIAL VERSION. THAT ONE
       DOESN'T SUFFER FROM LOG 0 PROBLEMS.

    .. math::
        L = -\\log p({label}_i|{output}_i)
          = \\log {label}_i! + \\exp({output}_i) - {output}_i {label}_i

    Output and label must have the same shape.

    Parameters
    ----------
    weight : float or None
        Global scalar weight for loss.
    batch_axis : int, default 0
        The axis that represents mini-batch.
    """
    def __init__(self, weight=None, batch_axis=0, **kwargs):
        super(Poisson, self).__init__(weight, batch_axis, **kwargs)

    def hybrid_forward(self, F, output, label, sample_weight=None):
        label = _reshape_label_as_output(F, output, label)
        loss = F.exp(output) - output * label
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)
```
label is not necessarily a constant. It needs to be included in the loss

### CODE REVIEW

**Comment by piiswrong** [comment]
cc @smolix @szha for reference
